// tslint-consistent-codestyle/rules/constParametersRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var tsutils_1 = require("tsutils");
var ts = require("typescript");
var Lint = require("tslint");
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithFunction(sourceFile, walk);
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
function walk(ctx) {
    tsutils_1.collectVariableUsage(ctx.sourceFile).forEach(function (variable, identifier) {
        if (!isParameter(identifier.parent) || !isConst(identifier, ctx.sourceFile))
            return;
        for (var _i = 0, _a = variable.uses; _i < _a.length; _i++) {
            var use = _a[_i];
            if (tsutils_1.isReassignmentTarget(use.location))
                ctx.addFailureAtNode(use.location, "Cannot reassign constant parameter '" + identifier.text + "'.");
        }
    });
}
function isParameter(node) {
    switch (node.kind) {
        case ts.SyntaxKind.Parameter:
            return true;
        case ts.SyntaxKind.BindingElement:
            return tsutils_1.getDeclarationOfBindingElement(node).kind === ts.SyntaxKind.Parameter;
        default:
            return false;
    }
}
function isConst(name, sourceFile) {
    if (name.parent.kind === ts.SyntaxKind.Parameter)
        return tsutils_1.getJsDoc(name.parent, sourceFile).some(jsDocContainsConst);
    return tsutils_1.parseJsDocOfNode(name, true, sourceFile).some(jsDocContainsConst);
}
function jsDocContainsConst(jsDoc) {
    if (jsDoc.tags !== undefined)
        for (var _i = 0, _a = jsDoc.tags; _i < _a.length; _i++) {
            var tag = _a[_i];
            if (tag.tagName.text === 'const' || tag.tagName.text === 'constant')
                return true;
        }
    return false;
}
