// tslint-consistent-codestyle/rules/parameterPropertiesRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var utils = require("tsutils");
var rules_1 = require("../src/rules");
var ALL_OR_NONE_OPTION = 'all-or-none';
var LEADING_OPTION = 'leading';
var TRAILING_OPTION = 'trailing';
var READONLY_OPTION = 'readonly';
var MEMBER_ACCESS_OPTION = 'member-access';
var ALL_OR_NONE_FAIL = 'don\'t mix parameter properties with regular parameters';
var LEADING_FAIL = 'parameter properties must precede regular parameters';
var TRAILING_FAIL = 'regular parameters must precede parameter properties';
var READONLY_FAIL = 'parameter property must be readonly';
var MEMBER_ACCESS_FAIL = 'parameter property must have access modifier';
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithWalker(new ParameterPropertyWalker(sourceFile, this.ruleName, {
            allOrNone: this.ruleArguments.indexOf(ALL_OR_NONE_OPTION) !== -1,
            leading: this.ruleArguments.indexOf(LEADING_OPTION) !== -1,
            trailing: this.ruleArguments.indexOf(TRAILING_OPTION) !== -1,
            readOnly: this.ruleArguments.indexOf(READONLY_OPTION) !== -1,
            memberAccess: this.ruleArguments.indexOf(MEMBER_ACCESS_OPTION) !== -1,
        }));
    };
    return Rule;
}(rules_1.AbstractConfigDependentRule));
exports.Rule = Rule;
var ParameterPropertyWalker = (function (_super) {
    tslib_1.__extends(ParameterPropertyWalker, _super);
    function ParameterPropertyWalker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ParameterPropertyWalker.prototype.walk = function (sourceFile) {
        var _this = this;
        var cb = function (node) {
            if (node.kind === ts.SyntaxKind.Constructor)
                _this._checkConstructorDeclaration(node);
            return ts.forEachChild(node, cb);
        };
        return ts.forEachChild(sourceFile, cb);
    };
    ParameterPropertyWalker.prototype._checkConstructorDeclaration = function (node) {
        var parameters = node.parameters;
        var length = parameters.length;
        if (length === 0)
            return;
        var index = -1;
        for (var i = 0; i < length; ++i) {
            if (utils.isParameterProperty(parameters[i])) {
                index = i;
                break;
            }
        }
        if (index === -1)
            return;
        if (this.options.allOrNone) {
            var start = parameters[0].getStart(this.getSourceFile());
            var end = parameters[parameters.length - 1].getEnd();
            if (index > 0) {
                this.addFailure(start, end, ALL_OR_NONE_FAIL);
            }
            else {
                for (var i = index + 1; i < length; ++i) {
                    if (!utils.isParameterProperty(parameters[i])) {
                        this.addFailure(start, end, ALL_OR_NONE_FAIL);
                        break;
                    }
                }
            }
        }
        else if (this.options.leading) {
            var regular = index > 0;
            for (var i = index; i < length; ++i) {
                if (utils.isParameterProperty(parameters[i])) {
                    if (regular)
                        this.addFailureAtNode(parameters[i], LEADING_FAIL);
                }
                else {
                    regular = true;
                }
            }
        }
        else if (this.options.trailing) {
            for (var i = index; i < length; ++i)
                if (!utils.isParameterProperty(parameters[i]))
                    this.addFailureAtNode(parameters[i], TRAILING_FAIL);
        }
        if (this.options.memberAccess) {
            for (var i = index; i < length; ++i) {
                var parameter = parameters[i];
                if (utils.isParameterProperty(parameter) && !utils.hasAccessModifier(parameter))
                    this.addFailureAtNode(parameter, MEMBER_ACCESS_FAIL);
            }
        }
        if (this.options.readOnly) {
            for (var i = index; i < length; ++i) {
                var parameter = parameters[i];
                if (utils.isParameterProperty(parameter) && !utils.hasModifier(parameter.modifiers, ts.SyntaxKind.ReadonlyKeyword))
                    this.addFailureAtNode(parameter, READONLY_FAIL);
            }
        }
    };
    return ParameterPropertyWalker;
}(Lint.AbstractWalker));
